### PR TITLE
Implement v2 of Pokemon API

### PR DESCRIPTION
**Pull request by jfberry**
This is a re-implementation of the pokemon retrieval API in an attempt to:
* Improve consistency and ease of use of API
* Remove reactmap specific workarounds
* Provide an equivalent (experimental) grpc implementation as a fore-runner to streaming/subscription support
* Retain the high performance of the v1 api

It is largely mygod's work from #134 with a slightly modified API, and re-factored code to allow the v1 api to exist in its original form until it can be deprecated.

Implementation details: The matching logic is as follows:

* All rulesets are categorized by Pokemon and form, the form is optional, or you can match against all pokemon
* If the current Pokemon matches a specific Pokemon-form, that ruleset will be used, if not found then an attempt will be made against Pokemon/all forms, otherwise the default (everything else) ruleset is used.
* A row is returned in the result if any rule in the ruleset is matched.
* A rule is matched if every single set standard is satisfied by the current row. (Therefore, overall it is an OR of AND's, and hence a Disjunctive Normal Form.)
* A consequence of this logic is that an empty ruleset will match nothing. To include everything, one could, for example, have a rule matching IV in range [-1, 100].

Path to merge:
* Working implementation somewhere (eg https://github.com/WatWowMap/ReactMap/pull/809 updated)
* Performance check against v1 on the same database to see performance is in line with previous solution (proving both performance adequate, and v1 implementation continues to behave as expected)
* GRPC does not need to be proven working, and will be considered experimental for now, and streaming support will absolutely be a seperate PR

### CODE REVIEW

File: decoder/api_pokemon_scan_v2.go [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+		pokemon int16
+		form    int16
+	}
+
+	dnfFilters := make(map[dnfFilterLookup][]ApiPokemonDnfFilter)
+
+	for _, filter := range retrieveParameters.DnfFilters {
+		if len(filter.Pokemon) > 0 {
+			for _, keyString := range filter.Pokemon {
+				var pokemonId int16 = -1
```
Comment (Mygod) [inline decoder/api_pokemon_scan_v2.go:62]: `keyString.Pokemon`?

**Comment by Mygod** [comment]
Updated https://github.com/WatWowMap/ReactMap/pull/809 but I think this PR is buggy and I conveniently cannot push to this branch.

File: decoder/api_pokemon_scan_v2.go (hunk 1 above)
Comment (jfberry) [reply decoder/api_pokemon_scan_v2.go:62]: i don't understand this observation

**Comment by jfberry** [comment]
> Updated [WatWowMap/ReactMap#809](https://github.com/WatWowMap/ReactMap/pull/809) but I think this PR is buggy and I conveniently cannot push to this branch.

You can submit a PR to this branch, no problem - I didn't want to push changes this extensive into your branch given that at least some of them may need to be reverted.
What problems are you experiencing?

File: decoder/api_pokemon_scan_v2.go (hunk 1 above)
Comment (Mygod) [reply decoder/api_pokemon_scan_v2.go:62]: Is this variable supposed to be -1 always? Then why not use a const?

Comment (jfberry) [reply decoder/api_pokemon_scan_v2.go]: Right, I get it now.  Codeblind!

**Comment by Mygod** [comment]
Seems working now. :+1:

File: decoder/api_pokemon_scan_v2.go [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+			filter.AtkIv != nil && (pokemonLookup.Atk < filter.AtkIv.Min || pokemonLookup.Atk > filter.AtkIv.Max) ||
+			filter.DefIv != nil && (pokemonLookup.Def < filter.DefIv.Min || pokemonLookup.Def > filter.DefIv.Max) ||
+			filter.Level != nil && (pokemonLookup.Level < filter.Level.Min || pokemonLookup.Level > filter.Level.Max) ||
+			filter.Cp != nil && (pokemonLookup.Cp < filter.Cp.Min || pokemonLookup.Cp > filter.Cp.Max) ||
+			filter.Gender != nil && (pokemonLookup.Gender < filter.Gender.Min || pokemonLookup.Gender > filter.Gender.Max) ||
+			filter.Size != nil && (pokemonLookup.Size < filter.Size.Min || pokemonLookup.Size > filter.Size.Max) {
+			return false
+		}
+
+		if pvpLookup == nil ||
```
Comment (Mygod) [inline decoder/api_pokemon_scan_v2.go:110]: Should be `pvpLookup != nil && (...)`

Comment (Mygod) [reply decoder/api_pokemon_scan_v2.go:110]: https://github.com/UnownHash/Golbat/pull/134/files#diff-599121fdc02df91e88c5bf7eceb2a14ce6e65b7312ce0d7c8559a0cb6921e3a6R627

Comment (Mygod) [reply decoder/api_pokemon_scan_v2.go:110]: Actually it should be
```go
        if filter.Little != nil && (pvpLookup == nil || pvpLookup.Little < filter.Little.Min || pvpLookup.Little > filter.Little.Max) ||
            filter.Great != nil && (pvpLookup == nil || pvpLookup.Great < filter.Great.Min || pvpLookup.Great > filter.Great.Max) ||
            filter.Ultra != nil && (pvpLookup == nil || pvpLookup.Ultra < filter.Ultra.Min || pvpLookup.Ultra > filter.Ultra.Max) {
            return false
        }
```

What's the point of having a separate `pvpLookup` object anyways? Completely unnecessary complexity. 🤷

Comment (jfberry) [reply decoder/api_pokemon_scan_v2.go]: Good catch, have pushed.
The pvpLookup object is rebuilt only when ohbem data is available (which is not every update). The core lookup data is rebuilt every time anything else changes.  
Originally it was more complex and generic - but through testing this was found to substantially impact performance with only a theoretical future benefit (if they introduce new leagues, for example).

**Comment by jfberry** [comment]
> Seems working now. 👍

Excellent. Now we just need to find someone with a big enough system to run a v1/v2 comparison so we can see if we have maintained the performance for some equivalent filters and dive in if not.

**Review by TurtIeSocks — APPROVED** [review]
## V1:
```
11|ne_golbat  | INFO 2023-10-06 14:46:58 GetPokemonInArea - total time 958.834748ms (locked time 359ns), 1071627 scanned, 261 skipped, 1398 returned
11|ne_golbat  | INFO 2023-10-06 14:47:01 GetPokemonInArea - total time 922.522647ms (locked time 514ns), 1071188 scanned, 290 skipped, 1398 returned
11|ne_golbat  | INFO 2023-10-06 14:47:03 GetPokemonInArea - total time 927.582359ms (locked time 5.557µs), 1071206 scanned, 294 skipped, 1399 returned
11|ne_golbat  | INFO 2023-10-06 14:47:05 GetPokemonInArea - total time 940.796403ms (locked time 891ns), 1071996 scanned, 280 skipped, 1400 returned
11|ne_golbat  | INFO 2023-10-06 14:47:06 GetPokemonInArea - total time 931.650198ms (locked time 489ns), 1071436 scanned, 249 skipped, 1397 returned
11|ne_golbat  | INFO 2023-10-06 14:47:08 GetPokemonInArea - total time 1.126422715s (locked time 803ns), 1071788 scanned, 333 skipped, 1397 returned
11|ne_golbat  | INFO 2023-10-06 14:47:10 GetPokemonInArea - total time 945.524921ms (locked time 784ns), 1072289 scanned, 261 skipped, 1401 returned
11|ne_golbat  | INFO 2023-10-06 14:47:12 GetPokemonInArea - total time 947.441633ms (locked time 576ns), 1072230 scanned, 291 skipped, 1401 returned
11|ne_golbat  | INFO 2023-10-06 14:47:15 GetPokemonInArea - total time 908.027364ms (locked time 684ns), 1071861 scanned, 279 skipped, 1399 returned
```
## V2
```
11|ne_golbat  | INFO 2023-10-06 14:51:26 GetPokemonInAreaV2 - scan time 903.173111ms (locked time 310ns), 1113463 scanned, 312 skipped, 1435 returned
11|ne_golbat  | INFO 2023-10-06 14:51:27 GetPokemonInAreaV2 - scan time 952.739043ms (locked time 820ns), 1158112 scanned, 297 skipped, 1412 returned
11|ne_golbat  | INFO 2023-10-06 14:51:28 GetPokemonInAreaV2 - scan time 916.690746ms (locked time 573ns), 1147860 scanned, 311 skipped, 1393 returned
11|ne_golbat  | INFO 2023-10-06 14:51:30 GetPokemonInAreaV2 - scan time 1.255027342s (locked time 379ns), 1045369 scanned, 326 skipped, 1275 returned
11|ne_golbat  | INFO 2023-10-06 14:51:31 GetPokemonInAreaV2 - scan time 862.24818ms (locked time 56.161µs), 1048146 scanned, 264 skipped, 1287 returned
11|ne_golbat  | INFO 2023-10-06 14:51:31 GetPokemonInAreaV2 - scan time 812.885507ms (locked time 521ns), 1048451 scanned, 244 skipped, 1287 returned
11|ne_golbat  | INFO 2023-10-06 14:51:32 GetPokemonInAreaV2 - scan time 881.712777ms (locked time 377ns), 1048381 scanned, 285 skipped, 1287 returned
11|ne_golbat  | INFO 2023-10-06 14:51:37 GetPokemonInAreaV2 - scan time 901.83577ms (locked time 555ns), 1067487 scanned, 224 skipped, 1316 returned
11|ne_golbat  | INFO 2023-10-06 14:51:38 GetPokemonInAreaV2 - scan time 891.550087ms (locked time 402ns), 1067382 scanned, 249 skipped, 1316 returned
11|ne_golbat  | INFO 2023-10-06 14:51:44 GetPokemonInAreaV2 - scan time 865.239609ms (locked time 462ns), 1073664 scanned, 274 skipped, 1320 returned
11|ne_golbat  | INFO 2023-10-06 14:51:45 GetPokemonInAreaV2 - scan time 919.12354ms (locked time 681ns), 1073744 scanned, 274 skipped, 1320 returned
11|ne_golbat  | INFO 2023-10-06 14:51:46 GetPokemonInAreaV2 - scan time 855.471754ms (locked time 460ns), 1076107 scanned, 245 skipped, 1321 returned
11|ne_golbat  | INFO 2023-10-06 14:51:47 GetPokemonInAreaV2 - scan time 856.611085ms (locked time 430ns), 1073233 scanned, 223 skipped, 1321 returned
11|ne_golbat  | INFO 2023-10-06 14:51:49 GetPokemonInAreaV2 - scan time 924.846841ms (locked time 404ns), 1076108 scanned, 288 skipped, 1320 returned
11|ne_golbat  | INFO 2023-10-06 14:51:49 GetPokemonInAreaV2 - scan time 828.504954ms (locked time 460ns), 1073134 scanned, 255 skipped, 1317 returned
11|ne_golbat  | INFO 2023-10-06 14:51:50 GetPokemonInAreaV2 - scan time 886.905386ms (locked time 831ns), 1074936 scanned, 229 skipped, 1317 returned
11|ne_golbat  | INFO 2023-10-06 14:51:51 GetPokemonInAreaV2 - scan time 934.347161ms (locked time 475ns), 1071534 scanned, 265 skipped, 1316 returned
11|ne_golbat  | INFO 2023-10-06 14:51:52 GetPokemonInAreaV2 - scan time 841.154314ms (locked time 714ns), 1071201 scanned, 265 skipped, 1315 returned
11|ne_golbat  | INFO 2023-10-06 14:51:53 GetPokemonInAreaV2 - scan time 866.662451ms (locked time 288ns), 1070246 scanned, 267 skipped, 1312 returned
11|ne_golbat  | INFO 2023-10-06 14:51:53 GetPokemonInAreaV2 - scan time 908.240747ms (locked time 771ns), 1069340 scanned, 255 skipped, 1313 returned
11|ne_golbat  | INFO 2023-10-06 14:51:54 GetPokemonInAreaV2 - scan time 1.135335738s (locked time 437ns), 1068604 scanned, 410 skipped, 1313 returned
11|ne_golbat  | INFO 2023-10-06 14:51:55 GetPokemonInAreaV2 - scan time 972.252754ms (locked time 654ns), 1068062 scanned, 267 skipped, 1312 returned
11|ne_golbat  | INFO 2023-10-06 14:51:56 GetPokemonInAreaV2 - scan time 960.784819ms (locked time 609ns), 1066433 scanned, 290 skipped, 1308 returned
11|ne_golbat  | INFO 2023-10-06 14:51:56 GetPokemonInAreaV2 - scan time 857.643795ms (locked time 438ns), 1065955 scanned, 257 skipped, 1306 returned
11|ne_golbat  | INFO 2023-10-06 14:51:57 GetPokemonInAreaV2 - scan time 915.861581ms (locked time 303ns), 1062496 scanned, 268 skipped, 1298 returned
11|ne_golbat  | INFO 2023-10-06 14:51:59 GetPokemonInAreaV2 - scan time 913.968317ms (locked time 635ns), 1063928 scanned, 297 skipped, 1306 returned
```

**Comment by jfberry** [comment]
> ## V1:
> ```
> 11|ne_golbat  | INFO 2023-10-06 14:46:58 GetPokemonInArea - total time 958.834748ms (locked time 359ns), 1071627 scanned, 261 skipped, 1398 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:01 GetPokemonInArea - total time 922.522647ms (locked time 514ns), 1071188 scanned, 290 skipped, 1398 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:03 GetPokemonInArea - total time 927.582359ms (locked time 5.557µs), 1071206 scanned, 294 skipped, 1399 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:05 GetPokemonInArea - total time 940.796403ms (locked time 891ns), 1071996 scanned, 280 skipped, 1400 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:06 GetPokemonInArea - total time 931.650198ms (locked time 489ns), 1071436 scanned, 249 skipped, 1397 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:08 GetPokemonInArea - total time 1.126422715s (locked time 803ns), 1071788 scanned, 333 skipped, 1397 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:10 GetPokemonInArea - total time 945.524921ms (locked time 784ns), 1072289 scanned, 261 skipped, 1401 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:12 GetPokemonInArea - total time 947.441633ms (locked time 576ns), 1072230 scanned, 291 skipped, 1401 returned
> 11|ne_golbat  | INFO 2023-10-06 14:47:15 GetPokemonInArea - total time 908.027364ms (locked time 684ns), 1071861 scanned, 279 skipped, 1399 returned
> ```
> 
> ## V2
> ```
> 11|ne_golbat  | INFO 2023-10-06 14:51:26 GetPokemonInAreaV2 - scan time 903.173111ms (locked time 310ns), 1113463 scanned, 312 skipped, 1435 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:27 GetPokemonInAreaV2 - scan time 952.739043ms (locked time 820ns), 1158112 scanned, 297 skipped, 1412 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:28 GetPokemonInAreaV2 - scan time 916.690746ms (locked time 573ns), 1147860 scanned, 311 skipped, 1393 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:30 GetPokemonInAreaV2 - scan time 1.255027342s (locked time 379ns), 1045369 scanned, 326 skipped, 1275 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:31 GetPokemonInAreaV2 - scan time 862.24818ms (locked time 56.161µs), 1048146 scanned, 264 skipped, 1287 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:31 GetPokemonInAreaV2 - scan time 812.885507ms (locked time 521ns), 1048451 scanned, 244 skipped, 1287 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:32 GetPokemonInAreaV2 - scan time 881.712777ms (locked time 377ns), 1048381 scanned, 285 skipped, 1287 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:37 GetPokemonInAreaV2 - scan time 901.83577ms (locked time 555ns), 1067487 scanned, 224 skipped, 1316 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:38 GetPokemonInAreaV2 - scan time 891.550087ms (locked time 402ns), 1067382 scanned, 249 skipped, 1316 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:44 GetPokemonInAreaV2 - scan time 865.239609ms (locked time 462ns), 1073664 scanned, 274 skipped, 1320 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:45 GetPokemonInAreaV2 - scan time 919.12354ms (locked time 681ns), 1073744 scanned, 274 skipped, 1320 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:46 GetPokemonInAreaV2 - scan time 855.471754ms (locked time 460ns), 1076107 scanned, 245 skipped, 1321 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:47 GetPokemonInAreaV2 - scan time 856.611085ms (locked time 430ns), 1073233 scanned, 223 skipped, 1321 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:49 GetPokemonInAreaV2 - scan time 924.846841ms (locked time 404ns), 1076108 scanned, 288 skipped, 1320 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:49 GetPokemonInAreaV2 - scan time 828.504954ms (locked time 460ns), 1073134 scanned, 255 skipped, 1317 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:50 GetPokemonInAreaV2 - scan time 886.905386ms (locked time 831ns), 1074936 scanned, 229 skipped, 1317 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:51 GetPokemonInAreaV2 - scan time 934.347161ms (locked time 475ns), 1071534 scanned, 265 skipped, 1316 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:52 GetPokemonInAreaV2 - scan time 841.154314ms (locked time 714ns), 1071201 scanned, 265 skipped, 1315 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:53 GetPokemonInAreaV2 - scan time 866.662451ms (locked time 288ns), 1070246 scanned, 267 skipped, 1312 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:53 GetPokemonInAreaV2 - scan time 908.240747ms (locked time 771ns), 1069340 scanned, 255 skipped, 1313 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:54 GetPokemonInAreaV2 - scan time 1.135335738s (locked time 437ns), 1068604 scanned, 410 skipped, 1313 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:55 GetPokemonInAreaV2 - scan time 972.252754ms (locked time 654ns), 1068062 scanned, 267 skipped, 1312 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:56 GetPokemonInAreaV2 - scan time 960.784819ms (locked time 609ns), 1066433 scanned, 290 skipped, 1308 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:56 GetPokemonInAreaV2 - scan time 857.643795ms (locked time 438ns), 1065955 scanned, 257 skipped, 1306 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:57 GetPokemonInAreaV2 - scan time 915.861581ms (locked time 303ns), 1062496 scanned, 268 skipped, 1298 returned
> 11|ne_golbat  | INFO 2023-10-06 14:51:59 GetPokemonInAreaV2 - scan time 913.968317ms (locked time 635ns), 1063928 scanned, 297 skipped, 1306 returned
> ```

This is excellent feedback. Can you just confirm a little about what you were matching on in each case? 
And I assume v1 is from this PR against a current develop/main branch of reactmap?

**Comment by TurtIeSocks** [comment]
> This is excellent feedback. Can you just confirm a little about what you were matching on in each case? And I assume v1 is from this PR against a current develop/main branch of reactmap?

Yes, these are both against the same Golbat. V1 on my prod map and V2 in my dev env. 

I was just using pretty standard global settings:
Screenshot 2023-10-06 at 11 03 20 AM

This is zoomed out to pretty unreasonable levels, as you can see, over a million records were searched.

**Comment by jfberry** [comment]
> I was just using pretty standard global settings: Screenshot 2023-10-06 at 11 03 20 AM
> 
> This is zoomed out to pretty unreasonable levels, as you can see, over a million records were searched.

Can you try a little bit more complex search that would

> > This is excellent feedback. Can you just confirm a little about what you were matching on in each case? And I assume v1 is from this PR against a current develop/main branch of reactmap?
> 
> Yes, these are both against the same Golbat. V1 on my prod map and V2 in my dev env.
> 
> I was just using pretty standard global settings: Screenshot 2023-10-06 at 11 03 20 AM
> 
> This is zoomed out to pretty unreasonable levels, as you can see, over a million records were searched.

This is quite a good search because the DNF would have multiple things to match (ie 0-0%, 100-100%, top1 of each of three leagues). 

Have you/mygod tested whether per pokemon advanced filtering is also working at the like-for-like performance levels?

**Pull request merged**